src/shacl/tokens.ts

```diff
@@ -0,0 +1,347 @@
+// name.)
+// This function is called by the SHACL parser. It is memoized because the
+// arguments are small and unlikely to change often, and the parser needs to be
+// fast, so we should avoid re-computing.
+export const getShaclTokenMap: (
+  prefixes: { shacl: string; xsd: string }
+) => TokenMap = memoize((prefixes: { shacl: string; xsd: string }) => {
+  const prefixWithShacl = makePrefixer(prefixes.shacl);
+  const prefixWithXsd = makePrefixer(prefixes.xsd);
+
+  // Add the prefixed local names to the SHACL token map now that we know the
+  // prefixes.
+  const shaclTokenMap = localNames.reduce((tokenMap, localName) => {
+    const tokenName = `SHACL_${localName}`;
+    const prefixedTokenName = `${tokenName}_prefixed`;
+
+    return {
+      ...tokenMap,
+      [prefixedTokenName]: createToken({
+        name: prefixedTokenName,
+        pattern: prefixWithShacl(localName),
+        categories: [tokenMap[tokenName], turtleTokenMap.PNAME_LN],
+      }),
+    };
+  }, shaclUnprefixedTokenMap);
+
+  // Add the prefixed local names to the XSD token map now that we know the
+  // prefixes.
+  return xsdLocalNames.reduce((tokenMap, localName) => {
+    const tokenName = `SHACL_xsd_${localName}`;
+    const prefixedTokenName = `${tokenName}_prefixed`;
+
+    return {
+      ...tokenMap,
+      [prefixedTokenName]: createToken({
+        name: prefixedTokenName,
+        pattern: `${prefixWithXsd(localName)}`,
+        categories: [tokenMap[tokenName], turtleTokenMap.PNAME_LN],
+      }),
+    };
+  }, shaclTokenMap);
+}, isDeepEqual);
+
+// Get the index of PNAME_NS and IRIREF so that we can re-use existing Turtle
+// tokens but ensure that our special SHACL/XSD tokens are inserted at the
+// right place (since order of tokens matters for chevrotain).
+const pnameIndex = turtleTokenTypes.indexOf(sparqlTokenMap.PNAME_NS);
+const iriIndex = turtleTokenTypes.indexOf(turtleTokenMap.IRIREF);
+
+// tokenMap keys will need to be sorted in reverse order so that tokens with
+// partial overlap are in the right order in the TokenType array.
+const reverseSort = (a, b) => {
+  // @ts-ignore: unused variable
+  const [aIgnore, aName, aRemainder] = a.split('_');
+  // @ts-ignore: unused variable
+  const [bIgnore, bName, bRemainder] = b.split('_');
+  if (aName === bName) {
+    if (aRemainder && bRemainder) {
+      return 0; // treat as lexicographically the same for sorting
+    }
+    return aRemainder ? 1 : -1;
+  }
+  return aName < bName ? 1 : bName < aName ? -1 : 0;
+};
+
+// Given SHACL and XSD prefixes, this method returns an array of Turtle +
+// SHACL/XSD tokens, including tokens for prefixed local names, with the
+// SHACL/XSD tokens inserted at the proper positions so that they are matched
+// before the more generic Turtle tokens.
+export const getShaclTokenTypes: (
+  prefixes: { shacl: string; xsd: string }
+) => TokenType[] = memoize((prefixes: { shacl: string; xsd: string }) => {
```

Comment:
is this called a bunch somewhere else? only see it once in this codebase

Reply:
Only here for now, though it's possible that other parsers in the future could make use of it (in the same way that we re-use the Turtle token types in various other places). I mentioned above the reasoning for the memoization (in addition to just defensively keeping the parser fast since this method does a decent amount of work).

Reply:
I see it's `memoize-one` now, so no worries about cache size. 👍